### ISSUE TITLE
2.4.11 Focus Appearance (Minimum) (AA) - clarification needed

### ISSUE THREAD

**Issue by MichellanneLi** [w3c/wcag#1341]
Regarding the new sc 2.4.11 Focus Appearance (Minimum) (AA), I'd like clarification around the change of contrast requirement: "Change of contrast: The color change for the focus indication area has a contrast ratio of at least 3:1 with the colors of the unfocused state." 

If a focus indicator is added that 1.) does not rely on color and 2.) meets minimum adjacent contrast requirements, is it really necessary for the indicator to also have a 3:1 contrast with the unfocused state? 

If so, this contradicts past precedent of allowing for indicators besides color. It will also severely restrict the colors designers are able to use for default and focus states.

If not, could some clarifying language be added to this sc? 

Also, browser defaults have historically be treated as the exception, as they are not dictated by content creators. Will that continue to be the case regarding focus indicators? If so, language around that exception should be added.

**Comment by alastc** [comment]
> this contradicts past precedent of allowing for indicators besides color.

What indicator doesn't rely on color in some way? If there is some visual change (e.g. addition of a line), that line will have a change of colour from the unfocused state, as the line replaces pixels of another colour.

For the contrast SCs (including text-contrast) to work, we have to differentiate hue and brightness. Contrast differences are based on differing levels of brightness. "Not relying on color" generally means not relying on hue, and the brightness difference (contrast) can be relied on.

Can you find (or think of) examples that would be both visible and restricted?

**Comment by patrickhlauke** [comment]
> What indicator doesn't rely on color in some way?

what if it adds a line AND changes text or something to bold instead? if a user can't perceive the line due to low contrast, but can perceive the change to bold, does that count?

**Comment by patrickhlauke** [comment]
> "Not relying on color" generally means not relying on hue, and the brightness difference (contrast) can be relied on.

as an aside, this old chestnut should now REALLY be clarified normatively somewhere...

**Comment by alastc** [comment]
> what if it adds a line AND changes text or something to bold instead? if a user can't perceive the line due to low contrast, but can perceive the change to bold, does that count?

If the line doesn't contrast with the non-focused state, it doesn't count towards the min focus area. That leaves you with the change of pixels from standard to bold text, and I *really* don't want to recommend that as a technique. I suspect it would not meet the size requirement in most cases, but it would be a pain to work out.

**Comment by patrickhlauke** [comment]
> I really don't want to recommend that as a technique

not necessarily about what we'd want to *recommend*, but rather what does and doesn't normatively pass/fail. admittedly, it'd be a pain to determine if the "boldness" of the text passes the area requirement (probably not in most cases) ... but it might?

**Comment by jared-w-smith** [comment]
I'm also very confused by this. When I read "...with the colors of the unfocused state" it's not clear what the "unfocused state" refers to. Is it the unfocused state of the "user interface component" or the "focus indication area" when the component is in an unfocused state? I initially interpreted this as the former (and think @MichellanneLi is too) and the Understanding document supports this, but I believe the intent is in fact the latter. If it is the latter, this is problematic because the "focus indication area" cannot have an "unfocused state" because the focus indicator itself is not focusable.

Assuming my interpretation is correct, a much more clear wording would be "The color change for the focus indication area has a contrast ratio of at least 3:1 with the colors of the focus indication area when the user interface component is in an unfocused state".

Consider a list of dark blue links on a white background in a sidebar that present a large, adjacent dark blue arrow when they are focused (and the links themselves do not change color or appearance). This blue arrows would have >3:1 contrast to the white "unfocused state" area they now occupy, but would have 1:1: contrast with the blue text of the "unfocused state" of the links. Would these pass or fail? I think they would pass, but this depends entirely on what "unfocused state" refers to - and this is not clear from the SC text.

**Comment by mraccess77** [comment]
I think @jared-w-smith recommended proposal is good to clear up any confusion.

**Comment by alastc** [comment]
I think Jared's change puts it back closer to a previous version. If folks think that is clearer, happy to use it. (I don't think I can tell anymore!)

**Comment by awkawk** [comment]
The questions and issues raised made me think about this SC more, and I’d like to suggest some additional changes.

1. I think that it is causing some problems that we are defining focus indication _area_ and not talking about mathematical area (length x width) in all cases. When we are talking about contrast we can’t be talking about area because it isn’t about area. Of course, we are meaning to use area in this definition to mean a set of pixels, so not strictly mathematical area, but then the first thing we do is ask for the minimum area. How about “Focus indicator” or “Focus indication region”? I prefer the former because it is familiar.
2. The definition is stateless in its current form. This is not a problem with the Minimum Area bullet because the only way to measure this is for the component to be focused, but it feels awkward for Change of Contrast and especially for Adjacent Contrast.
3. The Change of Contrast item is super hard to parse. I tried several iterations and have a new attempt below.
4. I made a change so that each part of the SC talks about the focus indicator and I think that this is easier to read as a result:
a.	SC text: change in defined term to focus indicator
b.	Minimum area: talks about “area of the focus indicator”
c.	Change of contrast: made into two sentences. Find the area of pixels that are 3:1. Compare to minimum area.
d.	Adjacent contrast: just changed to focus indicator but now it is unambiguous that it is talking about the focused state.
e.	Unobscured: changed “the item with focus” to “the focus indicator” since that is what this SC is about.

Leading to:

For the keyboard focus indicator of each User Interface Component, all of the following are true:
•	Minimum area: The area of the focus indicator is greater than or equal to a 1 CSS pixel border of the focused control, or has a thickness of at least 8 CSS pixels along the shortest side of the element.
•	Change of contrast: The focus indicator includes a combined total area of pixels which have a contrast ratio of at least 3:1 against the same pixels when the control is not focused. This area must be at least equal to the Minimum Area.
•	Adjacent contrast: The focus indicator has a contrast ratio of at least 3:1 against all adjacent colors for the minimum area or greater, or has a thickness of at least 2 CSS pixels.
•	Unobscured: The focus indicator is not entirely hidden by author-created content.

Focus Indicator:
The set of pixels show the focus(link) through a change in color when a User Interface Component is in the focused state.

Next best alternative for change of contrast:
•	Change of contrast: The focus indicator has a contrast ratio of at least 3:1 between the colors presented in the focus indicator and the same pixels when unfocused, for the minimum area or greater.

Thoughts?

**Comment by jared-w-smith** [comment]
I believe @awkawk's suggestions provide much improvement. Here are a few suggested tweaks:

> Minimum area: The area of the focus indicator is ~~greater than or equal to~~ _at least as large as_ the _area covered by_ a 1 CSS pixel _wide_ border of the focused control, or has a ~~thickness~~ _width_ of at least 8 CSS pixels along the shortest side of the element.

This resolves the distinct problem of equating an area to a perimeter. While "border" in CSS does have an area if it has a width >0 pixels, "border" in math typically means "perimeter" which does not have an area. This also, I think, clarifies that the focus indicator area doesn't have to be in the same location as the border. Perhaps there is a reason for using "thickness" instead of "width", but I believe the HTML/CSS term to define the size of things is much clearer here.

> Change of contrast: The focus indicator ~~includes a combined total area of pixels which have~~ _pixel area has_ a contrast ratio of at least 3:1 against the same pixels when the control is not focused. This _contrasting_ area must be at least ~~equal to~~ _as large as_ the Minimum Area.

Harmonizes with language above. Removes "equal to", which, as above, suggests "the exact same place as" when instead it means means "the same area as".

> Adjacent contrast: The focus indicator has a contrast ratio of at least 3:1 against all adjacent colors for _at least_ the minimum area~~ or greater~~, or has a thickness of at least 2 CSS pixels.

Harmony of "at least". I'm not particularly tied to this, but suggest either "at least" or "greater than or equal to" consistently.

> Focus Indicator: The ~~set of~~ pixels _that_ show the focus...

I think "set of" introduces an additional concept that is not necessary.

Finally, I suspect there is some long history that I'm unaware of regarding "not _entirely_ hidden", but I'm rather opposed to the notion than an element could be 99% obscured yet still pass this AA SC. The only reasonable solution I can think of is to require that "no part" be hidden as is currently in AAA.

**Comment by alastc** [comment]
I'll come back to the text-changes and try to conglomerate them, but on this:
> I suspect there is some long history that I'm unaware of regarding "not entirely hidden", but I'm rather opposed to the notion than an element could be 99% obscured yet still pass this AA SC.
There is quite a bit of history!

As it stands 2.4.7 doesn’t cover scenarios where the focus indicator is temporarily hidden by things like sticky headers, footers, or non-modal dialogues.

Focus appearance without that bullet would not (IMO) cover those temporary scenarios either. The focus is defined on the control and if that happens to be underneath other content because you need to scroll a bit further, that is not to do with the control as such.

When we discussed the issue in general, there is a desire to cover it completely, but when you look at the implementations it is a difficult area. There’s a lot of overlap with how user-agents deal with things, and it is hard to make the author responsible for all these. For example:

- Tabbing down the page (in Firefox at least) will often cut-off the bottom of the indicator;
- You can use scroll-padding to help keep the focus indicator in view, but it is a fixed value. If the user has text-spacing, text-sizing, or zoom which triggers word-wrap then a sticky footer would increase in height;
- Media carousels which show half an item at the end of a row for usability (to indicate there is more) would not show a full indicator.

I’m sure there are many more scenarios where it wouldn’t be under the author’s control, we’ve just scratched the surface. See also #1271 and #1304 for more discussion.

**Comment by jared-w-smith** [comment]
@alastc Your comments here are in reference to the focus indicator being obscured, but that is not what the SC text refers to. It instead indicates "The **item with focus** is not entirely hidden...". It's one thing to allow the focus indicator to be partially obscured at AA, but, as written, this SC allows the all but 1 pixel of the focusable element to be obscured, and that is, I believe, problematic. It seems odd that one could meet the Level A (Focus **Visible**) and Level AA (Focus **Appearance**) criteria when the item in question is allowed to be almost entirely **not visible** and **without appearance**).

**Comment by mbgower** [comment]
@jared-w-smith 
> It seems odd that one could meet the Level A (Focus Visible) and Level AA (Focus Appearance) criteria when the item in question is allowed to be almost entirely not visible and without appearance.

As it stands at AA, one doesn't have to have any part of it visible to meet Focus Visible, so ensuring it is not entirely hidden is an arguable (if unimpressive sounding) advancement.

IMO, it is achieves considerably more than this bare minimum. There is now the notion that it must be at least partially visible. All of the existing implementations of sticky foots, non-modal dialogs, etc., that currently fully obscure the item with focus would fail.

As well, in practical application, it increases the active consideration by authors of keeping the item with focus in the viewport.

Once the wording gets shaken out a bit more, the combination of 1) clearer contrast for focus indication and 2) the item with focus being within the viewport makes for a compelling increase in usability for users with low vision.

**Comment by mraccess77** [comment]
I think the nuance for 2.4.7 is that it can be not visible -- but it must be able to be made visible with navigation mechanisms - e.g. pressing up/down arrow or tabbing/shift tabbing, etc.  If it was truly not visible without requiring the mouse or dismissing a dialog then it would fail 2.4.7 -- at least that is how I understood what the group had agreed to.

**Comment by mbgower** [comment]
@mraccess77
> I think the nuance for 2.4.7 is that it can be not visible 

 right, the "a mode of operation where the keyboard focus indicator is visible."

**Comment by alastc** [comment]
@mbgower sorry, just saw this. For future, the answer is generally going to be yes!

**Comment by mbgower** [comment]
@alastc Okay. As won't surprise you, it took a LONG time to try to incorporate the ideas and still make it clear. I'm pretty happy with the PR update, although still feel like the change of contrast bullet is bizarrely tough to get across, given how obvious the desired outcome is!

**Comment by mbgower** [comment]
Note that I no longer use the term "focus indication area", and so that definition no longer needs to exist.
_Focus indicator_ already exists as a term in the Focus Visible SC, so I don't think we need to define it suddenly for 2.2 (having gone this long without it). However, if for some reason you wanted to, I think a better definition would be:

>Focus indicator
The pixels that change to visually indicate when a user interface component is in a focused state

**Comment by mbgower** [comment]
Pushed changes to the PR to include the use of "perimeter", along with its definition and  the above possible definition for focus indicator (and removal of the prior area dan)

**Comment by mbgower** [comment]
@alastc I couldn't actually see where the definitions were linked, so that is missing from this change.

**Comment by mbgower** [comment]
@alastc, the PR addresses this ticket and https://github.com/w3c/wcag/issues/1372.
Summary of issue and  status:
The ticket asked for clarification on Change of contrast (addressed in PR) as well as asking if there is an exception for browser defaults. (not addressed in PR, but easy to add!)
Minor discussion on another form of indicator other than visual, such as changes of text. (not addressed)
A suggestion that "Not relying on color" be defined as part of the normative text, as something like 'generally means not relying on hue, and the brightness difference (contrast) can be relied on.' (not addressed)
Concern with phrase "..with the colors of the unfocused state" (addressed)
Awk provided a [full rewrite ](https://github.com/w3c/wcag/issues/1341#issuecomment-680896275)pivoting on use of "focus indicator" (incorporated with mods)
Jared suggested tweaks of AWK's wording (incorporated with mods)

PR
- adopted https://github.com/w3c/wcag/issues/1372's intro to contain assessment to the time when the element receives focus (which resolves the assessment being carried out after the user has moved the point of regard, etc)
- altered wording of Minimum area so that the calculation is clearer
- altered wording of Change of contrast so that it is now talking about the pixels involved in the focus indicator
- attempted a definition of Focus indicator
- attempted a definition of Perimeter

**Comment by mbgower** [comment]
Draft AGWG response:
> Thank you for your issue. The SC wording has been updated to provide clarity around what is meant by 'Change of contrast'. Along with other SC changes, we hope the intent and steps necessary to meet this SC are now clearer.

>The user agent focus indicator can be used to meet Focus Visible; however if a user agent implementation fails to meet Focus Appearance (Minimum) requirements, the page would fail. OR
> An exception for the default user agent focus indicator has been added.

**Comment by mraccess77** [comment]
Which is it? "The default user agent can be used to meet the Level A requirement, Focus Visible." OR "An exception for the default user agent focus indicator has been added."  Are we saying that it will be one of the two pending a vote on this subject by the Accessibility Guidelines working group?

**Comment by mbgower** [comment]
@mraccess77 
> Are we saying that it will be one of the two pending a vote on this subject by the Accessibility Guidelines working group?

I'm saying that there is no language about user agent exemption in the current wording. I'm flagging that the WG needs to discuss this, and confirm a decision, which will result in a choice of which wording goes into the response to this issue.

**Comment by alastc** [comment]
Oops, undoing auto-close.

**Comment by alastc** [comment]
The things left to do for this issue were:
- Create an update for the understanding document about the default focus indicator not being enough;
- Create an update for the SC providing an exception for the browser default.

(I.e. opposing outcomes to put to the group.)

**Comment by mraccess77** [comment]
If we accept the default focus indicator then we'd have to accept the caret of an input or text area unless we don't accept those situations.

**Comment by alastc** [comment]
In the last meeting we agreed to NOT add an exception for the default indicator, and merged the paragraph about default indicators in the understanding doc. https://www.w3.org/2020/11/10-ag-minutes.html#item16

**Comment by patrickhlauke** [comment]
Coming in late on this, but: if browser defaults are now not exempted, a pass/fail will be dependent on what the browser defaults are. If a browser changes their default, a past result may become outdated. From a testing perspective, this will also require that all (common?) browsers will need to be used for testing, since they may have different defaults.

**Comment by alastc** [comment]
> if browser defaults are now not exempted

They had never been excepted in this proposed SC, in fact some of the initial impetus was the poor defaults.

Past results I see, but I think that would apply to other SC as well? 

In terms of testing the defaults, as several of the defaults won't pass it's a case of setting an indicator, at least until the defaults improve.

**Comment by patrickhlauke** [comment]
> They had never been excepted in this proposed SC, in fact some of the initial impetus was the poor defaults.

ah my bad, had the non-text contrast SC in mind.
